server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedSortedSetDocValuesSyntheticFieldLoader.java

```diff
@@ -110,8 +155,65 @@ public int count() {
         }
 
         @Override
-        public void write(XContentBuilder b) throws IOException {
-            this.writer.write(b);
+        public SortedSetDocValues getValues() {
+            return dv;
+        }
+    }
+
+    private static class DocValuesWithIgnoredSortedKeyedValues implements FlattenedFieldSyntheticWriterHelper.SortedKeyedValues {
+        private final FlattenedFieldSyntheticWriterHelper.SortedKeyedValues docValues;
+        private final TreeSet<BytesRef> ignoredValues;
+
+        private BytesRef currentFromDocValues;
+
+        private DocValuesWithIgnoredSortedKeyedValues(
+            FlattenedFieldSyntheticWriterHelper.SortedKeyedValues docValues,
+            TreeSet<BytesRef> ignoredValues
+        ) {
+            this.docValues = docValues;
+            this.ignoredValues = ignoredValues;
+        }
+
+        @Override
+        public BytesRef next() throws IOException {
```

Comment:
Nit: let's add some high-level comments below so that we remember the logic in a year from now.